src/class-model.ts

```diff
@@ -82,40 +80,19 @@ class BaseClassModel {
     return extend(this, props);
   }
 
+  /** Properties set in the fast instantiator compiled constructor, included here for type information */
+  [$readOnly]!: true;
+  [$type]!: IClassModelType<TypesForModelPropsDeclaration<any>>;
   /** @hidden */
   readonly [$env]?: any;
   /** @hidden */
   readonly [$parent]?: IStateTreeNode | null;
   /** @hidden */
-  [$memos] = null;
+  [$identifier]?: any;
   /** @hidden */
-  [$memoizedKeys] = null;
+  [$memos]!: Record<string, any> | null;
   /** @hidden */
-  [$identifier]?: any;
-
-  constructor(
-    snapshot: InputsForModel<InputTypesForModelProps<TypesForModelPropsDeclaration<any>>> | undefined,
-    context: InstantiateContext,
-    parent: IStateTreeNode | null,
-    /** @hidden */ hackyPreventInitialization = false
-  ) {
-    if (hackyPreventInitialization) {
-      return;
-    }
-
-    this[$env] = context.env;
-    this[$parent] = parent;
-
-    (this.constructor as IClassModelType<any>)[$fastInstantiator](this as any, snapshot, context);
```

Comment:
This call was megamorphic before -- all the MQT classes subclass this one, and then this class ends up calling the instantiator. Because they all share the same parent class, this callsite is shared among all of em!

Reply:
So simple, but so effective!